src/psvm/Kernel.chpl

```diff
@@ -1,9 +1,62 @@
+use Documents;
+
 enum KernelType {LINEAR, POLYNOMIAL, GAUSSIAN, LAPLACIAN};
+use KernelType;
 
+// TODO
 class Kernel {
 
-  // TODO
-  // methods..
+  proc calcKernel(a: Sample, b: Sample): real {
+    select this.kernelType {
+      when LINEAR {
+        return innerProduct(a, b) / sqrt(a.twoNormSq * b.twoNormSq);
+      }
+      when POLYNOMIAL {
+        halt('KernelType not supported');
+      }
+      when GAUSSIAN {
+        halt('KernelType not supported');
+      }
+      when LAPLACIAN {
+        halt('KernelType not supported');
+      }
+      otherwise {
+        halt('KernelType not supported');
+      }
+    }
+  }
+
+  proc calcKernelWithLabel(a: Sample, b: Sample) {
+    var result = calcKernel(a, b);
+    if a.classLabel != b.classLabel then result *= -1.0;
+    return result;
+  }
+
+  proc innerProduct(a: Sample, b: Sample) {
+    var norm: real;
+
+    // TODO: Refactor as iterator:
```

Comment:
Looks like a good exercise for me.

Reply:
Possibly! I'm happy to help you with that, if you're interested.

Reply:
Would you mind pushing this branch?